### ISSUE TITLE
🐛 [AutoFixture.Xunit2] MemberAutoDataAttribute generates fixtures for test method parameters provided from MemberData

### ISSUE THREAD

**Issue by JelleKerkstra** [AutoFixture/AutoFixture#1407]
### Describe the Bug
When I use the MemberAutoDataAttribute to provide data for some of my test method parameters and the AutoDataAttribute to create the remaining parameters, the AutoDataAttribute will also generate data for the parameters provided by the member data. The types of the arguments provided by the member data, are recursive data structures which I don't want to be created by the fixture. Because my tests are written around many of those structures (and in all cases they are provided by either member data or inline data), I want to avoid the tedious task of creating customizations for each of those types to prevent generating data for their properties that will cause recursion.

(The above also goes for the InlineAutoDataAttribute)

### Scenario
```
public sealed class AutoFakeItEasyDataAttribute : AutoDataAttribute
{
}

public sealed class MemberAutoFakeItEasyDataAttribute : MemberAutoDataAttribute
{
    public MemberAutoFakeItEasyDataAttribute(string memberName, params object[] values)
        : base(new AutoFakeItEasyDataAttribute(), memberName, values)
    {
    }
}


public interface IService { }

public class RecursiveDataStructure
{
    public RecursiveDataStructure Data { get; set; }
}

public class Test
{
    public static IEnumerable<object[]> Test_MemberData
    {
        get
        {
            yield return new object[] { new RecursiveDataStructure { Data = new RecursiveDataStructure() } };
        }
    }

    [Theory, MemberAutoFakeItEasyData(nameof(Test_MemberData))]
    public void TestMethod(RecursiveDataStructure data, IService service)
    {
        Assert.NotNull(data);
    }
}
```

The test above will throw the following exception: 
```
  Message: 
AutoFixture.ObjectCreationExceptionWithPath : AutoFixture was unable to create an instance of type AutoFixture.Kernel.SeededRequest because the traversed object graph contains a circular reference. Information about the circular path follows below. This is the correct behavior when a Fixture is equipped with a ThrowingRecursionBehavior, which is the default. This ensures that you are being made aware of circular references in your code. Your first reaction should be to redesign your API in order to get rid of all circular references. However, if this is not possible (most likely because parts or all of the API is delivered by a third party), you can replace this default behavior with a different behavior: on the Fixture instance, remove the ThrowingRecursionBehavior from Fixture.Behaviors, and instead add an instance of OmitOnRecursionBehavior:

fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList()
    .ForEach(b => fixture.Behaviors.Remove(b));
fixture.Behaviors.Add(new OmitOnRecursionBehavior());

                

Request path:
	Sample.RecursiveDataStructure Data
	  Sample.RecursiveDataStructure
	    Sample.RecursiveDataStructure data


  Stack Trace: 
ThrowingRecursionHandler.HandleRecursiveRequest(Object request, IEnumerable`1 recordedRequests)
RecursionGuard.HandleRecursiveRequest(Object request)
ObsoletedMemberShims.RecursionGuard_HandleRecursiveRequest(RecursionGuard guard, Object request)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
SeedIgnoringRelay.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
Postprocessor`1.Create(Object request, ISpecimenContext context)
AutoPropertiesTarget.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
TerminatingWithPathSpecimenBuilder.Create(Object request, ISpecimenContext context)
BehaviorRoot.Create(Object request, ISpecimenContext context)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
PropertyRequestRelay.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
Postprocessor`1.Create(Object request, ISpecimenContext context)
AutoPropertiesTarget.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
TerminatingWithPathSpecimenBuilder.Create(Object request, ISpecimenContext context)
BehaviorRoot.Create(Object request, ISpecimenContext context)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
AutoPropertiesCommand`1.Execute(Object specimen, ISpecimenContext context)
Postprocessor`1.Create(Object request, ISpecimenContext context)
AutoPropertiesTarget.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
TerminatingWithPathSpecimenBuilder.Create(Object request, ISpecimenContext context)
BehaviorRoot.Create(Object request, ISpecimenContext context)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
SeedIgnoringRelay.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
Postprocessor`1.Create(Object request, ISpecimenContext context)
AutoPropertiesTarget.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
TerminatingWithPathSpecimenBuilder.Create(Object request, ISpecimenContext context)
BehaviorRoot.Create(Object request, ISpecimenContext context)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
ParameterRequestRelay.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
Postprocessor`1.Create(Object request, ISpecimenContext context)
AutoPropertiesTarget.Create(Object request, ISpecimenContext context)
CompositeSpecimenBuilder.Create(Object request, ISpecimenContext context)
TerminatingWithPathSpecimenBuilder.Create(Object request, ISpecimenContext context)
BehaviorRoot.Create(Object request, ISpecimenContext context)
RecursionGuard.Create(Object request, ISpecimenContext context)
Fixture.Create(Object request, ISpecimenContext context)
SpecimenContext.Resolve(Object request)
AutoDataAttribute.Resolve(ParameterInfo p)
AutoDataAttribute.GetData(MethodInfo testMethod)
<>c__DisplayClass5_0.<GetData>b__0(DataAttribute attr)
<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
SelectArrayIterator`2.ToList()
EnumerableExtensions.Zip[T,TResult](IEnumerable`1 sequences, Func`2 resultSelector)+MoveNext()
```

### Expected Behavior
What I would expect, is that AutoFixture would not try to create an instance of the ```RecursiveDataStructure``` since it is provided by the member data.

### More Information
For me the following approach solved this issue as the cause for this behavior lies within the ```AutoDataAttribute``` and the ```CompositeDataAttribute```. I am curious to know if the current behavior of those attributes I am experiencing is intended (and if so, for what reason).

```
public sealed class AutoFakeItEasyDataAttribute : DataAttribute
{
    private readonly Lazy<IFixture> fixtureLazy;

    public AutoFakeItEasyDataAttribute()
    {
        fixtureLazy = new Lazy<IFixture>(() => new Fixture().Customize(new AutoFakeItEasyCustomization()));
    }

    public override IEnumerable<object[]> GetData(MethodInfo testMethod)
    {
        if (testMethod == null) throw new ArgumentNullException(nameof(testMethod));

        return new[] { GetData(testMethod.GetParameters()).ToArray() };
    }

    internal IEnumerable<object> GetData(IEnumerable<ParameterInfo> parameters)
    {
        foreach (var p in parameters)
        {
            CustomizeFixture(p);
            yield return Resolve(p);
        }
    }

    private void CustomizeFixture(ParameterInfo p)
    {
        var customizeAttributes = p.GetCustomAttributes()
            .OfType<IParameterCustomizationSource>()
            .OrderBy(x => x, new CustomizeAttributeComparer());

        foreach (var ca in customizeAttributes)
        {
            var c = ca.GetCustomization(p);
            fixtureLazy.Value.Customize(c);
        }
    }

    private object Resolve(ParameterInfo p)
        => new SpecimenContext(fixtureLazy.Value).Resolve(p);

    private class CustomizeAttributeComparer : Comparer<IParameterCustomizationSource>
    {
        public override int Compare(IParameterCustomizationSource x, IParameterCustomizationSource y)
        {
            var xfrozen = x is FrozenAttribute;
            var yfrozen = y is FrozenAttribute;

            if (xfrozen && !yfrozen)
            {
                return 1;
            }

            if (yfrozen && !xfrozen)
            {
                return -1;
            }

            return 0;
        }
    }
}

public class CompositeDataAttribute : DataAttribute
{
    private readonly AutoFakeItEasyDataAttribute autoFakeItEasyDataAttribute = new();
    private readonly DataAttribute dataAttribute;

    public CompositeDataAttribute(DataAttribute dataAttribute)
    {
        this.dataAttribute = dataAttribute;
    }

    public override IEnumerable<object[]> GetData(MethodInfo testMethod)
    {
        var data = dataAttribute.GetData(testMethod).ToList();

        return data.Select(set =>
            set.Concat(autoFakeItEasyDataAttribute.GetData(testMethod.GetParameters().Skip(set.Length))).ToArray());
    }
}

public sealed class MemberAutoFakeItEasyDataAttribute : CompositeDataAttribute
{
    public MemberAutoFakeItEasyDataAttribute(string memberName, params object[] parameters)
        : base(new MemberDataAttribute(memberName, parameters))
    {
    }
}
```

Thanks!

**Comment by JelleKerkstra** [comment]
I ran into a related issue today where the number of test cases provided for by the member data did not match the cases run by XUnit. I managed to reproduce the issue in the unit test below. This might help with implementing behavior for the original issue (while it might be an issue on its own)

```
using System;
using System.Collections.Generic;
using System.Linq;
using AutoFixture;
using AutoFixture.AutoFakeItEasy;
using AutoFixture.Xunit2;
using Xunit;

namespace Tests;

public class CompositeAttributeTests
{
    [Fact]
    public void Should_Return_Two_Test_Cases_From_Two_Partial_ArgumentSets()
    {
        var attribute = new MemberDataAutoFakeItEasyDataAttribute(nameof(Data));
        var data = attribute.GetData(typeof(CompositeAttributeTests).GetMethod(nameof(Test_Method))).ToList();

        Assert.Equal(2, data.Count);

        Assert.Equal(2, data[0].Length);
        Assert.Equal(data[0][0], 1);
        Assert.IsType<string>(data[0][1]);

        Assert.Equal(2, data[1].Length);
        Assert.Equal(data[1][0], 1);
        Assert.IsType<string>(data[1][1]);
    }

    public static IEnumerable<object[]> Data
    {
        get
        {
            yield return new object[] { 1 };
            yield return new object[] { 2 };
        }
    }

    public void Test_Method(int a, string b)
    {
        throw new NotImplementedException();
    }

    public sealed class MemberDataAutoFakeItEasyDataAttribute : MemberAutoDataAttribute
    {
        public MemberDataAutoFakeItEasyDataAttribute(string memberName, params object[] parameters) : base(
            new AutoFakeItEasyDataAttribute(), memberName, parameters)
        {
        }
    }

    public sealed class AutoFakeItEasyDataAttribute : AutoDataAttribute
    {
        public AutoFakeItEasyDataAttribute() : base(() => new Fixture().Customize(new AutoFakeItEasyCustomization()))
        {
        }
    }
}
```

**Comment by aivascu** [comment]
This is a very old and well-known issue #1142.

All AutoData attributes have been completely reworked for v5, fixing this bug and adding freezing injected arguments.
I was hoping to finally merge it over the holidays, but I've been doing OT for the past 3 weeks.

I've published a feature preview package `5.0.0-dataattributes0001`. If anyone wishes to help please test the package.
I can't tell anyone to use this version production but if you're stuck you're free to do so. I'll leave it listed until it will be merged into the main v5 branch.

Hope this helps, this is as much as I can do for now.

**Comment by aivascu** [comment]
Duplicate of #1142
